### DISCUSSION TITLE
What exactly happens after renderMathInElement() changes the HTML?

### DISCUSSION THREAD

**Discussion by Kombat-Batyanya** [KaTeX/KaTeX#2989] [Q&A]
Hi,
I am trying to figure out if I will be able to use KaTeX in TIBCO Spotfire JSViz visualizations. For those who don't know what Spotfire is, it is a corporate BI platform that can use embedded Chromium to render graphic elements written in JavaScript (among many other nice things) .

As far as I can see my renderMathInElement() call successfully parses LaTeX content in my DIV and changes the HTML code, but when it tries to render the result on the screen it gets stuck and renders nothing. No errors are thrown. 
This is my DIV just before and after the renderMathInElement() call: 

![my Spotfire JSViz DIV with KaTeX content](https://user-images.githubusercontent.com/2218736/117350950-229f2e80-ae7b-11eb-9f9f-fc1f5494c625.jpg)

_https://drive.google.com/file/d/18q0dhtSqpAijTs1OkB3bhhZ204SyWJjv/view?usp=sharing_

Hence the subject question, if the HTML is successfully replaced by renderMathInElement(), what could prevent Spotfire's Chromium to render it? Am I merely missing some additional libraries not disclosed in _https://katex.org/docs/autorender.html_ ?
For example, what does this url _http://www.w3.org/1998/Math/MathML_ do in the replaced code?

Thank you in advance.

**Comment by Kombat-Batyanya** [comment]
This issue seems to be a consequence of KaTeX not being able to render on SVG as described in _https://github.com/KaTeX/KaTeX/issues/375_

I managed to display a KaTeX content in a Spotfire JSViz that doesn't use SVGs.
Looks like I should try MathJax then.

**Reply by edemaine** [reply]
KaTeX output indeed requires an HTML renderer, not just an SVG renderer.  Does that answer your question?

**Reply by Kombat-Batyanya** [reply]
Thank you Erik for a quick response. Highly appreciated.
You can safely assume I have zero knowledge of modern html and javascript coding.
All I want is a nice chart on my dashboard that I can only achieve with LaTeX data labels.

I don't understand your distinction between HTML and SVG renderer . I followed the KaTeX documentation meticulously and included necessary js and css libraries and a call to renderMathInElement(document.body).

When my Spotfire JSViz only contains text, I can see LaTeX formulas on the screen.

When it contains graphics where LaTeX content is placed in text elements inside the SVG, such text label simply disappears from the rendered chart. 

Chrome Inspect Element confirms that my LaTeX content was replaced by a hierarchy of katex-labeled span's, so renderMathInElement performed at least a part of its duties successfully.

However, given that the six-year old issue https://github.com/KaTeX/KaTeX/issues/375 is still open I understand that KaTeX is currently not able to render what I need. Is this correct?

Regards, Dmitry

PS: I opened a separate discussion on Spotfire community to understand which application causes the problem.
_https://community.tibco.com/questions/latex-jsviz_

**Reply by edemaine** [reply]
That doesn't sound right. If you're able to render some KaTeX content, then you should be able to render all KaTeX content. #375 is about something else. If things are disappearing, I imagine it might be from a crash -- did you check the JavaScript console (if possible)?  Can you post an example of a formula that's failing to render?  Did you test it on the https://katex.org demo?

**Reply by Kombat-Batyanya** [reply]
The console is available and it is virgin clean from errors both when I see KaTeX formula on the screen and when I don't.
I did test the formula works in katex.org in the first place.  I also saw your recommendations in other posts to replace single backslashes in a formula with double ones. I tried both ways to no avail. 
I tried many different formulas, for example \\(\\int_0^1 xdx.\\) that I simply took from one of your _stackoverflow_ posts.

**Reply by Kombat-Batyanya** [reply]
I tried MathJax and I face exactly the same behaviour. I can render LaTeX formulas when there are no SVG tags in my script, and just an empty space when relevant data labels are inside an SVG container. So it seems the problem is on Spotfire side.

PS: I think the integral sign in KaTeX formulas is somewhat taller than in MathJax ones.

**Reply by edemaine** [reply]
Unfortunately, it's pretty hard to tell what's going wrong from here.  It sounds like you have a Chromium instance with the correct DOM that's not rendering that DOM. I don't *think* it has anything to do with KaTeX, but perhaps something else specific to this environment.

Can you try saving the generated page as HTML (in Element inspector, right click on the root and use Copy outerHTML) and opening the file in regular Chrome?  If that doesn't work, it would be useful to post the HTML (as a gist, for example). If it does work, it's further confirmation that it's specific to your environment and not KaTeX.

Hope this helps!

**Reply by Kombat-Batyanya** [reply]
Erik, 
Thank you for very interesting hints.

When I followed your suggestion, the generated HTML file shows the entire Spotfire interface with working buttons etc. but a greyed-out area of my chart bearing a note "it may have been moved, edited or deleted".

Investigating the issue further, the DOM that I obtained earlier by calling _console.log(document.body)_ immediately after _renderMathInElement(document.body)_ is clearly not the same DOM that is being rendered on my screen. Spotfire adds quite a bunch of its own stuff (menus, buttons, area dividers) on top of it and then includes the chart DOM in an iframe inside it. The html inside the _iframe_ seems to keep all the _katex_ spans as I shared in my original post.

So the question is whether _KaTeX_ is supposed to work from within iframes.

I am thinking on a potential workaround implemented by Dr.M.Bourne in _https://github.com/KaTeX/KaTeX/issues/375#issuecomment-152699930_  but that would be a pain in the neck for me as I only started looking at JavaScript code a couple of weeks ago regardless of my 35 year amateur coding experience.

Regards
Dmitry

**Reply by edemaine** [reply]
Oops, looking again at your original screenshot, I see now that you're trying to render a formula that's inside a `<text>` element of an SVG.  That won't work; KaTeX only supports math inside HTML.  You *might* be able to get it to work by wrapping the KaTeX output in an SVG `<foreignObject>` but this hasn't been tested.

It *might* also be possible with MathJax using its SVG rendering mode, but not without significant additional code.  ([This project](https://github.com/edemaine/cocreate) does [something along those lines](https://github.com/edemaine/cocreate/blob/main/client/RenderObjects.coffee), but it wouldn't be easy to adapt to your setting.)

**Reply by Kombat-Batyanya** [reply]
Thank you, Erik, for such a quick diagnosis of my situation. I am impressed. 
I will park the topic now and will try to approximate the desired outcome with simple multiple-line text elements.

In the modern copy-paste-link approach to software development "significant coding" is not encouraged. 

All the best,
Dmitry

**Reply by Kombat-Batyanya** [reply]
Just to confirm that SVG \<foreignObject\> approach works with KaTeX as per the example found [here](https://markmap.js.org/repl):

_https://drive.google.com/file/d/1iLOt7U05SY2NtyphwU7XE4SPR9Iz_Wil/view?usp=sharing_

Will the default documented call to renderMathInElement() automatically replace text inside \<foreignObject\> or do I need to allow that in options?

**Reply by edemaine** [reply]
Cool! It should find them; the [default `ignoredClasses`](https://katex.org/docs/autorender.html) does not list `foreignObject`.
